# Relatório de Análise V

## Tratamento de Dados Faltantes

In [1]:
import pandas as pd

In [3]:
dados = pd.read_csv('dados/aluguel_residencial.csv', sep = ';')

dados.head()

Tipo           Bairro  Quartos  Vagas  Suites  Area   Valor  \
0    Quitinete       Copacabana        1      0       0    40  1700.0   
1         Casa  Jardim Botânico        2      0       1   100  7000.0   
2  Apartamento           Centro        1      0       0    15   800.0   
3  Apartamento     Higienópolis        1      0       0    48   800.0   
4  Apartamento     Vista Alegre        3      1       0    70  1200.0   

   Condominio  IPTU  
0       500.0  60.0  
1         NaN   NaN  
2       390.0  20.0  
3       230.0   NaN  
4         NaN   NaN

In [10]:
# Retorna um dataframe com booleanos, True se for nulo e False se tiver dados
dados.isnull()

Tipo  Bairro  Quartos  Vagas  Suites   Area  Valor  Condominio   IPTU
0      False   False    False  False   False  False  False       False  False
1      False   False    False  False   False  False  False        True   True
2      False   False    False  False   False  False  False       False  False
3      False   False    False  False   False  False  False       False   True
4      False   False    False  False   False  False  False        True   True
...      ...     ...      ...    ...     ...    ...    ...         ...    ...
22575  False   False    False  False   False  False  False       False  False
22576  False   False    False  False   False  False  False       False  False
22577  False   False    False  False   False  False  False       False  False
22578  False   False    False  False   False  False  False       False  False
22579  False   False    False  False   False  False  False       False   True

[22580 rows x 9 columns]

In [15]:
# Retorna um dataframe com booleanos, False se for nulo e True se tiver dados
dados.notnull()

Tipo  Bairro  Quartos  Vagas  Suites  Area  Valor  Condominio   IPTU
0      True    True     True   True    True  True   True        True   True
1      True    True     True   True    True  True   True       False  False
2      True    True     True   True    True  True   True        True   True
3      True    True     True   True    True  True   True        True  False
4      True    True     True   True    True  True   True       False  False
...     ...     ...      ...    ...     ...   ...    ...         ...    ...
22575  True    True     True   True    True  True   True        True   True
22576  True    True     True   True    True  True   True        True   True
22577  True    True     True   True    True  True   True        True   True
22578  True    True     True   True    True  True   True        True   True
22579  True    True     True   True    True  True   True        True  False

[22580 rows x 9 columns]

In [12]:
# Monstra informações sobre os dados
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22580 entries, 0 to 22579
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Tipo        22580 non-null  object 
 1   Bairro      22580 non-null  object 
 2   Quartos     22580 non-null  int64  
 3   Vagas       22580 non-null  int64  
 4   Suites      22580 non-null  int64  
 5   Area        22580 non-null  int64  
 6   Valor       22571 non-null  float64
 7   Condominio  20765 non-null  float64
 8   IPTU        15795 non-null  float64
dtypes: float64(3), int64(4), object(2)
memory usage: 1.6+ MB


In [14]:
# Retorna os registros com nulo no campo 'Valor'.
dados[dados['Valor'].isnull()]

Tipo           Bairro  Quartos  Vagas  Suites  Area  Valor  \
58    Apartamento  Barra da Tijuca        2      1       1    70    NaN   
1492  Apartamento             Leme        2      0       0    75    NaN   
1683         Casa     Campo Grande        3      4       3   363    NaN   
2012  Apartamento         Botafogo        2      0       0    95    NaN   
2034  Apartamento       Copacabana        2      0       0    72    NaN   
4941         Casa     Campo Grande        3      2       1   100    NaN   
8568  Apartamento             Leme        2      0       1    75    NaN   
8947  Apartamento           Glória        3      0       1   135    NaN   
9149  Apartamento            Gávea        3      1       1   105    NaN   

      Condominio   IPTU  
58         970.0   68.0  
1492       878.0    NaN  
1683         NaN    NaN  
2012      1010.0  170.0  
2034       850.0    NaN  
4941         NaN    NaN  
8568       878.0    NaN  
8947       910.0  228.0  
9149       880.0  221.0

Como o que nos importa para estas análises são os valores dos aluguéis, vamos deletar os registros que não tem valor. 

Para isso usamos a função **.dropna()**.

In [17]:
# Removendo registros que não tem valor.
A = dados.shape[0]
dados.dropna(subset=['Valor'], inplace=True)
B = dados.shape[0]

#Total de registros excluídos.
A - B

9

In [20]:
dados[dados['Valor'].isnull()]

Empty DataFrame
Columns: [Tipo, Bairro, Quartos, Vagas, Suites, Area, Valor, Condominio, IPTU]
Index: []

## Tratamento de Dados Faltantes (Continuação)

In [23]:
dados[dados['Condominio'].isnull()]

Tipo                    Bairro  Quartos  Vagas  Suites  \
1                    Casa           Jardim Botânico        2      0       1   
4             Apartamento              Vista Alegre        3      1       0   
6      Casa de Condomínio           Barra da Tijuca        5      4       5   
7      Casa de Condomínio                     Ramos        2      2       0   
8             Apartamento                    Centro        1      0       0   
...                   ...                       ...      ...    ...     ...   
22510                Casa              Campo Grande        1      0       0   
22516        Casa de Vila                    Pavuna        2      0       0   
22535                Casa  Recreio dos Bandeirantes        4      2       2   
22547                Casa             Bento Ribeiro        2      1       0   
22558                Casa                    Centro        5      0       1   

       Area    Valor  Condominio  IPTU  
1       100   7000.0         NaN   NaN  
4        70   1200.0         NaN   NaN  
6       750  22000.0         NaN   NaN  
7        65   1000.0         NaN   NaN  
8        36   1200.0         NaN   NaN  
...     ...      ...         ...   ...  
22510    50    600.0         NaN   NaN  
22516    55    600.0         NaN   NaN  
22535   352   6500.0         NaN   NaN  
22547    60   1400.0         NaN   NaN  
22558   160   5000.0         NaN  80.0  

[1813 rows x 9 columns]

In [27]:
selecao = (dados['Tipo'] == 'Apartamento') & (dados['Condominio'].isnull())

dados[selecao]

Tipo           Bairro  Quartos  Vagas  Suites  Area   Valor  \
4      Apartamento     Vista Alegre        3      1       0    70  1200.0   
8      Apartamento           Centro        1      0       0    36  1200.0   
83     Apartamento       Praça Seca        2      0       0    80  1000.0   
117    Apartamento    Rocha Miranda        2      1       0    50  1100.0   
133    Apartamento  Barra da Tijuca        1      1       1    65  2350.0   
...            ...              ...      ...    ...     ...   ...     ...   
22328  Apartamento       Copacabana        3      0       0   130  4600.0   
22356  Apartamento        Guadalupe        1      0       0    40   550.0   
22373  Apartamento       Bonsucesso        1      0       0    60  1000.0   
22403  Apartamento         Flamengo        2      0       0    59  3500.0   
22474  Apartamento        Cascadura        3      0       1    70  1200.0   

       Condominio  IPTU  
4             NaN   NaN  
8             NaN   NaN  
83            NaN   NaN  
117           NaN   NaN  
133           NaN   NaN  
...           ...   ...  
22328         NaN   NaN  
22356         NaN   NaN  
22373         NaN   NaN  
22403         NaN   NaN  
22474         NaN   NaN  

[745 rows x 9 columns]

Quando temos dados boleanos uma forma de inverter o valor é usando o 'tio' **~** conforme exemplo a seguir.

In [32]:
# Removendo registros que são nulos
A = dados.shape[0]
dados = dados[~selecao]
B = dados.shape[0]

#Total de registros excluídos.
A - B

745

In [34]:
dados[dados['Condominio'].isnull()].shape[0]

1068

Como temos ainda 1068 registros nulos mas que não podemos excluir vamos colocar um valor padrão. Neste caso será 0.

In [37]:
dados.fillna(0, inplace=True)

In [38]:
dados[dados['Condominio'].isnull()].shape[0]

0

In [40]:
#Outra Maneira de fazer para multiplos atributos é:
dados = dados.fillna({'Condominio': 0, 'IPTU': 0})

In [42]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21826 entries, 0 to 22579
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Tipo        21826 non-null  object 
 1   Bairro      21826 non-null  object 
 2   Quartos     21826 non-null  int64  
 3   Vagas       21826 non-null  int64  
 4   Suites      21826 non-null  int64  
 5   Area        21826 non-null  int64  
 6   Valor       21826 non-null  float64
 7   Condominio  21826 non-null  float64
 8   IPTU        21826 non-null  float64
dtypes: float64(3), int64(4), object(2)
memory usage: 1.7+ MB


In [43]:
# Substituindo dados antigos pelos tratados.
dados.to_csv('dados/aluguel_residencial.csv', sep = ';', index=False)